packages/sdk/src/cross-chain-messenger.ts

```diff
@@ -1212,29 +1260,85 @@ export class CrossChainMessenger {
       throw new Error(`cannot get a state root for an L1 to L2 message`)
     }
 
-    // Try to find the output index that corresponds to the block number attached to the message.
-    // We'll explicitly handle "cannot get output" errors as a null return value, but anything else
-    // needs to get thrown. Might need to revisit this in the future to be a little more robust
-    // when connected to RPCs that don't return nice error messages.
+    let proposal: any
     let l2OutputIndex: BigNumber
-    try {
-      l2OutputIndex =
-        await this.contracts.l1.L2OutputOracle.getL2OutputIndexAfter(
-          resolved.blockNumber
+    if (await this.fpac()) {
+      // Get the respected game type from the portal.
+      const gameType =
+        await this.contracts.l1.OptimismPortal2.respectedGameType()
+
+      // Get the total game count from the DisputeGameFactory since that will give us the end of
+      // the array that we're searching over. We'll then use that to find the latest games.
+      const gameCount = await this.contracts.l1.DisputeGameFactory.gameCount()
+
+      // Find the latest 100 games (or as many as we can up to 100).
+      const latestGames =
+        await this.contracts.l1.DisputeGameFactory.findLatestGames(
+          gameType,
+          gameCount.sub(1),
+          Math.min(100, gameCount.toNumber())
         )
-    } catch (err) {
-      if (err.message.includes('L2OutputOracle: cannot get output')) {
+
+      // Find a game with a block number that is greater than or equal to the block number that the
+      // message was included in. We can use this proposal to prove the message to the portal.
+      const match = latestGames
```

Comment:
this doesn't really need to be in 2 iterations, could be done in a for/of loop which i think is way easier to read (just looks like c) but this looks like it works

Reply:
Changed to be a for loop